core/include/ikos/core/domain/memory/value.hpp

```diff
@@ -1162,17 +1187,115 @@ class ValueDomain final
+      }
+    }
+    if (read_mask == initialized_coverage) {
+      // The read bits are completely initialized
+      return ikos::core::Uninitialized::initialized();
+    } else if (read_mask == uninitialized_coverage) {
+      // The read bits are all uninitialized
+      return ikos::core::Uninitialized::uninitialized();
+    }
+    // The bits may or may not be initialized, so top.
+    return ikos::core::Uninitialized::top();
+  }
+
   /// \brief Create a new cell for a read
   VariableRef read_realize_single_cell(MemoryLocationRef base,
                                        const MachineInt& offset,
                                        const MachineInt& size,
                                        Signedness sign) {
+    ikos::core::Uninitialized addr_uninit =
+        is_read_uninitialized(base, offset, size);
+
     VariableRef new_cell = this->make_cell(base, offset, size, sign);
+    this->_scalar.uninit_refine(new_cell, addr_uninit);
     CellSetT cells = this->_cells.get(base);
-    cells.add(new_cell);
-    this->_cells.set(base, cells);
 
-    // TODO(marthaud): perform further reduction in case of partial overlaps
+    if (cells.is_empty()) {
+      // No cell found for the base address
+      this->_cells.set(base, CellSetT{new_cell});
+      return new_cell;
+    }
+
+    CellSetT new_cells = cells;
+    bool found = false;
+
+    IntInterval new_interval = this->cell_range(new_cell);
+    MachineInt new_lb = new_interval.lb();
+    MachineInt new_ub = new_interval.ub();
+
+    // Remove overlapping cells if uninitialized.
+    for (VariableRef cell : cells) {
+      if (cell == new_cell) {
+        found = true;
+      } else if (this->cell_overlap(cell, new_cell)) {
+        if (this->_scalar.uninit_is_uninitialized(cell)) {
+          // Make new uninitialized cells for the parts not covered
+          // by the new_cell.
+          IntInterval un_interval = this->cell_range(cell);
+          MachineInt un_lb = un_interval.lb();
+          MachineInt un_ub = un_interval.ub();
+          if (un_lb < new_lb) {
+            MachineInt low_size = new_lb - un_lb;
+            VariableRef low_cell = this->make_cell(base, un_lb, low_size, sign);
+            this->_scalar.uninit_refine(low_cell, Uninitialized::uninitialized());
+            new_cells.add(low_cell);
+          }
+          if (new_ub < un_ub) {
+            MachineInt high_size = un_ub - new_ub;
+            auto one = MachineInt(1, high_size.bit_width(), Unsigned);
+            MachineInt high_lb = new_ub + one;
+            VariableRef high_cell = this->make_cell(base, high_lb, high_size, sign);
+            this->_scalar.uninit_refine(high_cell, Uninitialized::uninitialized());
+            new_cells.add(high_cell);
+          }
+          this->_scalar.dynamic_forget(cell);
+          new_cells.remove(cell);
+        } else {
+          // Cell is at least potentially initialized, so keep it.
+        }
+      }
+    }
```

Comment:
Genuine question: do we really need to do this? does that increase the precision in some cases? I would think doing this for writes is enough. I think we can keep overlapping cells during reads.

Reply:
I agree this is unnecessary. Will remove.